### PR TITLE
Adaptive substepping

### LINKED ISSUE #23105: Add option to substep if a return mapping solve fails
## Reason
When a return mapping solve fails (e.g. in LaROMance) we cut the global timestep. Let's try if we can instead substep the failed solve, doubling the amount of substeps successively.

## Design
Add parameter to enable automatic substepping for failing return mapping solves.

## Impact
Might increase robustness/performance.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #23105

### CODE REVIEW

**Comment by dschwen** [comment]
This is not ready for merging, but I'd appreciate some feedback. Things left to do:

- [ ] Updating all inputs
- [ ] Adding this capability to the AD specialization

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C [hunk 1]
```diff
@@ -30,6 +30,26 @@ RadialReturnStressUpdateTempl<is_ad>::validParams()
       "effective_inelastic_strain_name",
       "Name of the material property that stores the effective inelastic strain");
   params.addParam<bool>("use_substep", false, "Whether to use substepping");
+
+  MooseEnum substeppingType("NONE ADAPTIVE ERROR_BASED INCREMENT_BASED", "NONE");
+  substeppingType.addDocumentation("NONE", "Do not use substepping");
+  substeppingType.addDocumentation(
+      "ADAPTIVE",
+      "Use adaptive substepping, where teh number of substeps is successively doubled until the "
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:38]:
```suggestion
      "Use adaptive substepping, where the number of substeps is successively doubled until the "
```

File: modules/tensor_mechanics/test/tests/substepping/ad_power_law_creep.i [hunk 2]
```diff
@@ -80,7 +80,6 @@
     activation_energy = 0.0
     temperature = temp
     # options for using substepping
-    use_substep = true
```
Comment (recuero) [inline modules/tensor_mechanics/test/tests/substepping/ad_power_law_creep.i:83]: Should this be `use_substepping = ERROR_BASED`?

File: modules/tensor_mechanics/test/tests/substepping/power_law_creep.i [hunk 3]
```diff
@@ -80,7 +80,6 @@
     activation_energy = 0.0
     temperature = temp
     # options for using substepping
-    use_substep = true
```
Comment (recuero) [inline modules/tensor_mechanics/test/tests/substepping/power_law_creep.i:83]: Maybe also here.

File: modules/tensor_mechanics/test/tests/substepping/ad_power_law_creep.i (hunk 2 above)
Comment (recuero) [reply modules/tensor_mechanics/test/tests/substepping/ad_power_law_creep.i:83]: Never mind, that should be taken care of by `tests`

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C [hunk 4]
```diff
@@ -30,6 +30,27 @@ RadialReturnStressUpdateTempl<is_ad>::validParams()
...
+  params.addParam<MooseEnum>(
+      "use_substepping", substeppingType, "Whether and how to use substepping");
+  params.addParam<bool>(
+      "adaptive_substepping",
+      false,
+      "Use adaptive substepping, where the number of substeps is successively doubled until the "
+      "return mapping model successfully converges or the maximum number of substeps is reached. ");
+
+  params.addDeprecatedParam<bool>(
+      "use_substep", false, "Whether to use substepping", "Use `use_substepping` instead");
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:53]: Are we doing the deadline thing to remove deprecated parameters? I cannot see a reason to keep this one for long.

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C [hunk 5]
```diff
@@ -385,12 +451,64 @@ RadialReturnStressUpdateTempl<false>::updateStateSubstep(RankTwoTensor & strain_
...
+    }
+    catch (MooseException & e)
+    {
+      // if we are not using adaptive substepping we just rethrow the exception
+      if (!_adaptive_substepping)
+        throw e;
+
+      // otherwise we double the number of substeps and try again
+      num_substeps *= 2;
+      if (num_substeps <= _maximum_number_substeps)
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:496]: We would effectively never reach anywhere closer to the maximum number of substeps. Not a big deal, but the user may think the maximum number of substeps was, at some point, reached. I'd leave it as it is though.

File: modules/tensor_mechanics/test/tests/substepping/tests [hunk 6]
```diff
@@ -5,45 +5,52 @@
...
     exodiff = 'power_law_substepping_out.e'
-    requirement = 'The system shall converge under large deformation owing to the help of '
-                  'substepping'
+    requirement = 'The system shall converge under large deformation owing to the help of substepping'
+  []
+  [power_law_adaptive_substepping]
+    # We only check that the simulation converges with adaptive substepping
+    type = 'RunApp'
+    input = 'power_law_creep.i'
+    cli_args = "Materials/power_law_creep/use_substepping=INCREMENT_BASED "
```
Comment (recuero) [inline modules/tensor_mechanics/test/tests/substepping/tests:41]: No `ERROR_BASED` and adaptive substepping?

**Review by recuero — COMMENTED** [review]
Looks good to me.

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C (hunk 5 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:446]: > We would effectively never reach anywhere closer to the maximum number of substeps.

Why is that?

File: modules/tensor_mechanics/test/tests/substepping/tests (hunk 6 above)
Comment (dschwen) [reply modules/tensor_mechanics/test/tests/substepping/tests:41]: I'll add one, also an AD test

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C (hunk 5 above)
Comment (recuero) [reply modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:446]: Maybe I am missing something according to your reply. The way I read it is: User selects maximum number of substeps = 7. Then, we have 1 substeps, failure, two substeps, failure, four substeps, failure, we try to make 8 substeps, but it's not less than or equal to 7 so we got too many substeps. The user wanted 7 and got 4.

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:446]: Oh, I see. I could try with the maximum number, but that could result in case where - if the user selects a maximum of 9 substeps, the algorithm tries 1,2,4,8,9, with 9 unlikely to be a substantial improvement. And the user really doesn't have a way to anticipate either scenario, as the suggested error based number of substeps may be 3, resulting in 3,6,7 substeps in your example.

I don't have a good answer here, other than maybe rewording the doc to state "...will double the amount of substeps as long as the number of substeps stays below the user specified maximum ..."

Comment (recuero) [reply modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:446]: Yeah, maybe just add that clarifying sentence to the doc.

**Review by recuero — APPROVED** [review]
